<?xml version="1.0"?>
<robot xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor"
       xmlns:controller="http://playerstage.sourceforge.net/gazebo/xmlschema/#controller"
       xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface"
       xmlns:xacro="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface">

  <property name="M_PI" value="3.1415926535897931" />

  <macro name="m3_neck" params="parent *origin">

     <joint name="head_j1" type="revolute">
      <axis xyz="0 0 1"/>
     <limit effort="100" velocity="${75*M_PI/180}" lower="${-56.0*M_PI/180}" upper="${56.0*M_PI/180}" />    
      <insert_block name="origin" /> <!-- transform from parent link to this joint frame -->
      <parent link="${parent}"/>
      <child link="neck_pan"/>
      <child link="xtion"/>
    </joint>
    
    <link name="neck_pan">
		<xacro:inertial_cuboid_with_pose mass="0.1" x_length="0.1" y_length="0.1" z_length="0.05">
			<origin xyz="0 0 0.0" rpy="0 0 0" />
		</xacro:inertial_cuboid_with_pose>

      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="neck_tilt_geom">		
		<!--<box size="0.01 0.01 0.01" />-->
		<!--<mesh filename="package://${MESH}/S2R1/S2R1_SH1_neckTilt.dae" />-->
		<mesh filename="package://${MESH}/S3R1/neck.stl" />
	      </geometry>  
      </visual>
      <!-- meshes  
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="neck_tilt_geom_collison">
		      <mesh filename="package://${MESH}/S3R1/neck.stl" />
	      </geometry>
      </collision>
      -->  
      <!-- primitive only -->
      <collision>
        <origin xyz="0.0 0 0.00833" rpy="0.0 0.0 0.0 " />
        <geometry>
          <cylinder radius="0.034" length="0.0166" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0.0 0 0.0473" rpy="0.0 0.0 0.0 " />
        <geometry>
          <cylinder radius="0.03" length="0.0614" />
        </geometry>
      </collision>
    </link>

    <joint name="head_j0" type="revolute">
      <axis xyz="0 -1 0" />
      <origin xyz="0 0 0.1103" rpy="0 0.314159265 0" />
      <parent link="neck_pan"/>
      <child link="shell"/>
      <limit effort="100" velocity="${20.0*M_PI/180}" lower="${-40.0*M_PI/180}" upper="${2.0*M_PI/180}" />
    </joint>
    
    <link name="shell">
		<xacro:inertial_cuboid_with_pose mass="1.0" x_length="0.2" y_length="0.2" z_length="0.1">
			<origin xyz="0 0 0.0" rpy="0 0 0" />
		</xacro:inertial_cuboid_with_pose>

      <visual>
	      <material name="GreyB" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="shell_geom">
		      <mesh filename="package://${MESH}/S3R1/S3.dae" />
	      </geometry>  
      </visual>
      <!-- meshes
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="S3_collision">
		      <box size="0.01 0.01 0.01" />
	      </geometry>
      </collision>
      -->  
      <!-- primitive only -->
      <collision>
        <origin xyz="0.0 0 0.0" rpy="1.5708 0.0  0.0 " />
        <geometry>
          <cylinder radius="0.0335" length="0.09" />
        </geometry>
      </collision>
      <collision>
        <origin xyz="0.005 0 0.07" rpy="0.0 0.0 0.0 " />
        <geometry>
          <box size="0.222 0.2 0.09" />
        </geometry>
      </collision>
    </link>
	
     <joint name="head_xtion_dummy" type="fixed">
	<origin xyz="0.0865 0.0095 0.0542" rpy="-1.49304175558102 0.0465121191942754 -1.5830006092006217"/>
      <axis xyz="0 0 1"/>
      <parent link="shell"/>
      <child link="xtion_link_rot"/>
    </joint>
    
    <link name="xtion_link_rot">
	<parent link="shell"/> 
		<xacro:inertial_cuboid_with_pose mass="0.1" x_length="0.1" y_length="0.1" z_length="0.05">
			<origin xyz="0 0 0.0" rpy="0 0 0" />
		</xacro:inertial_cuboid_with_pose>
    </link>  

     <joint name="head_xtion_dummy2" type="fixed">
	<limit effort="100" velocity="${75*M_PI/180}" lower="${-56.0*M_PI/180}" upper="${56.0*M_PI/180}" />
	<origin xyz="-0.04 0 0" rpy="0 -1.5708 1.5708"/>
      <axis xyz="0 0 1"/>
      <parent link="xtion_link_rot"/>
      <child link="xtion_link"/>
    </joint>

    <link name="xtion_link">
	<parent link="xtion_link_rot"/>      
		<xacro:inertial_cuboid_with_pose mass="0.1" x_length="0.1" y_length="0.1" z_length="0.05">
			<origin xyz="0 0 0.0" rpy="0 0 0" />
		</xacro:inertial_cuboid_with_pose>
    </link>
    
    <!--joint name="head_j2" type="revolute">
      <axis xyz="0 -1 0" /> 
      <origin xyz="0.07975 0 0.123536" rpy="0 0 0" />
      <parent link="shell"/>
	<child link="BB2"/>
      <limit effort="100" velocity="${20.0*M_PI/180}" lower="${-40.0*M_PI/180}" upper="${2.0*M_PI/180}" />
    </joint>
    
    <link name="BB2">
      <inertial>
        <mass value="0.409" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>

      <visual>
	      <material name="GreyB" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="shell_geom">
		      <mesh filename="package://${MESH}/S3R1/BB2.stl" />
	      </geometry>  
      </visual>
      
      <collision>
	      <origin xyz="-0.0141 0 0.0268" rpy="0.0 0.0 0.0 " />
	      <geometry name="S3_collision">
		      <box size="0.0474 0.157 0.037" />
	      </geometry>
      </collision>
    </link> -->
  </macro>

  <macro name="m3_head" params="parent *origin">
    <m3_neck parent="${parent}">
      <insert_block name="origin" />
    </m3_neck>
  </macro>


</robot>

